gcc/rust/typecheck/rust-hir-type-check.cc

```diff
@@ -123,44 +123,25 @@ TypeResolution::Resolve (HIR::Crate &crate)
 void
 TypeCheckExpr::visit (HIR::BlockExpr &expr)
 {
-  TyTy::BaseType *block_tyty
-    = new TyTy::TupleType (expr.get_mappings ().get_hirid ());
-
   expr.iterate_stmts ([&] (HIR::Stmt *s) mutable -> bool {
-    bool is_final_stmt = expr.is_final_stmt (s);
-    bool has_final_expr = expr.has_expr () && expr.tail_expr_reachable ();
-    bool stmt_is_final_expr = is_final_stmt && !has_final_expr;
-
     auto resolved = TypeCheckStmt::Resolve (s, inside_loop);
     if (resolved == nullptr)
       {
 	rust_error_at (s->get_locus_slow (), "failure to resolve type");
 	return false;
       }
 
-    if (stmt_is_final_expr)
-      {
-	delete block_tyty;
-	block_tyty = resolved;
-      }
-    else if (!resolved->is_unit ())
-      {
-	rust_error_at (s->get_locus_slow (), "expected () got %s",
-		       resolved->as_string ().c_str ());
-      }
-
     return true;
   });
 
   if (expr.has_expr ())
-    {
-      delete block_tyty;
-
-      block_tyty
-	= TypeCheckExpr::Resolve (expr.get_final_expr ().get (), inside_loop);
-    }
-
-  infered = block_tyty->clone ();
+    infered
+      = TypeCheckExpr::Resolve (expr.get_final_expr ().get (), inside_loop)
+	  ->clone ();
+  else if (expr.is_tail_reachable ())
+    infered = new TyTy::TupleType (expr.get_mappings ().get_hirid ());
+  else
+    infered = new TyTy::NeverType (expr.get_mappings ().get_hirid ());
```

Comment:
are we saying the end of every block is a NeverType now?

Reply:
It says if the block's tail is unreachable (for example, there is a `return`/`continue`/`break`), its type is `!`, otherwise its type is `()`. I don't think there is a problem.

Reply:
This is abusing the `!` type while assuming that the type analysis works in a specific way. This is too much technical debt IMO.

Reply:
For example, rustc can compile this. I don't think it is abusing the `!` type?
```rust
fn foo() -> ! {
    loop {};
    let mut a = 1;
    a = a + a;
}
```

Reply:
Yes, but only because the types unify successfully in the unreachable code. If the unreachable code produces an error the compilation will fail.

Reply:
I think there is nothing to do with this PR. After all, the problem already exists before this PR.